File: MoonSimulation/Renderers/FaceRenderer.cs
using SkiaSharp;
using MoonSimulation.Models;

namespace MoonSimulation.Renderers;

/// <summary>
/// Draws cute cartoon faces on celestial bodies.
/// </summary>
public static class FaceRenderer
{
    /// <summary>
    /// Returns a face expression name based on moon phase angle.
    /// Uses PhaseConstants for consistent angle thresholds.
    /// </summary>
    public static string GetMoonFaceExpression(double moonAngleDeg)
    {
        double angle = ((moonAngleDeg % 360) + 360) % 360;
        return angle switch
        {
            < PhaseConstants.NewMoonEnd => "sleepy",
            < PhaseConstants.WaxingCrescentEnd => "content",
            < PhaseConstants.FirstQuarterEnd => "happy",
            < PhaseConstants.WaxingGibbousEnd => "happy",
            < PhaseConstants.FullMoonEnd => "wow",
            < PhaseConstants.WaningGibbousEnd => "content",
            < PhaseConstants.LastQuarterEnd => "happy",
            < PhaseConstants.WaningCrescentEnd => "sleepy",
            _ => "sleepy"
        };
    }

    /// <summary>
    /// Draws a cute cartoon face on a celestial body.
    /// </summary>
    public static void DrawCuteFace(SKCanvas canvas, SKPoint center, float radius,
        string expression)
    {
        float eyeY = center.Y - radius * 0.12f;
        float eyeSpacing = radius * 0.25f;
        float eyeRadius = radius * 0.07f;

        using var eyePaint = new SKPaint
        {
            IsAntialias = true,
            Color = new SKColor(40, 40, 40, 200)
        };
        using var eyeHighlight = new SKPaint
        {
            IsAntialias = true,
            Color = new SKColor(255, 255, 255, 200)
        };
        using var mouthPaint = new SKPaint
        {
            IsAntialias = true,
            Style = SKPaintStyle.Stroke,
            StrokeWidth = radius * 0.04f,
            StrokeCap = SKStrokeCap.Round,
            Color = new SKColor(40, 40, 40, 180)
        };
        using var blushPaint = new SKPaint
        {
            IsAntialias = true,
            Color = new SKColor(255, 130, 130, 50)
        };

        switch (expression)
        {
            case "happy":
                DrawHappyFace(canvas, center, radius, eyeY, eyeSpacing, eyeRadius,
                    eyePaint, eyeHighlight, mouthPaint, blushPaint);
                break;
            case "sleepy":
                DrawSleepyFace(canvas, center, radius, eyeY, eyeSpacing, eyeRadius,
                    mouthPaint);
                break;
            case "wow":
                DrawWowFace(canvas, center, radius, eyeY, eyeSpacing, eyeRadius,
                    eyePaint, eyeHighlight, mouthPaint, blushPaint);
                break;
            case "content":
                DrawContentFace(canvas, center, radius, eyeY, eyeSpacing, eyeRadius,
                    eyePaint, eyeHighlight, mouthPaint);
                break;
        }
    }

    private static void DrawHappyFace(SKCanvas canvas, SKPoint center, float radius,
        float eyeY, float eyeSpacing, float eyeRadius,
        SKPaint eyePaint, SKPaint eyeHighlight, SKPaint mouthPaint, SKPaint blushPaint)
    {
        // Eyes
        canvas.DrawCircle(center.X - eyeSpacing, eyeY, eyeRadius, eyePaint);
        canvas.DrawCircle(center.X + eyeSpacing, eyeY, eyeRadius, eyePaint);
        // Eye highlights
        canvas.DrawCircle(center.X - eyeSpacing + eyeRadius * 0.3f, eyeY - eyeRadius * 0.3f,
            eyeRadius * 0.35f, eyeHighlight);
        canvas.DrawCircle(center.X + eyeSpacing + eyeRadius * 0.3f, eyeY - eyeRadius * 0.3f,
            eyeRadius * 0.35f, eyeHighlight);
        // Smile
        using var smilePath = new SKPath();
        smilePath.MoveTo(center.X - radius * 0.2f, center.Y + radius * 0.1f);
        smilePath.QuadTo(center.X, center.Y + radius * 0.3f,
            center.X + radius * 0.2f, center.Y + radius * 0.1f);
        canvas.DrawPath(smilePath, mouthPaint);
        // Blush cheeks
        canvas.DrawCircle(center.X - radius * 0.35f, center.Y + radius * 0.08f,
            radius * 0.1f, blushPaint);
        canvas.DrawCircle(center.X + radius * 0.35f, center.Y + radius * 0.08f,
            radius * 0.1f, blushPaint);
    }

    private static void DrawSleepyFace(SKCanvas canvas, SKPoint center, float radius,
        float eyeY, float eyeSpacing, float eyeRadius, SKPaint mouthPaint)
    {
        // Closed eyes (arcs)
        using (var closedEyePath = new SKPath())
        {
            closedEyePath.MoveTo(center.X - eyeSpacing - eyeRadius, eyeY);
            closedEyePath.QuadTo(center.X - eyeSpacing, eyeY + eyeRadius * 0.8f,
                center.X - eyeSpacing + eyeRadius, eyeY);
            canvas.DrawPath(closedEyePath, mouthPaint);
        }
        using (var closedEyePath2 = new SKPath())
        {
            closedEyePath2.MoveTo(center.X + eyeSpacing - eyeRadius, eyeY);
            closedEyePath2.QuadTo(center.X + eyeSpacing, eyeY + eyeRadius * 0.8f,
                center.X + eyeSpacing + eyeRadius, eyeY);
            canvas.DrawPath(closedEyePath2, mouthPaint);
        }
        // Gentle smile
        using (var sleepSmile = new SKPath())
        {
            sleepSmile.MoveTo(center.X - radius * 0.1f, center.Y + radius * 0.15f);
            sleepSmile.QuadTo(center.X, center.Y + radius * 0.22f,
                center.X + radius * 0.1f, center.Y + radius * 0.15f);
            canvas.DrawPath(sleepSmile, mouthPaint);
        }
        // Zzz
        SphereRenderer.DrawLabel(canvas, "z z", new SKPoint(center.X + radius * 0.5f, center.Y - radius * 0.3f),
            radius * 0.2f, new SKColor(200, 200, 255, 150));
    }

    private static void DrawWowFace(SKCanvas canvas, SKPoint center, float radius,
        float eyeY, float eyeSpacing, float eyeRadius,
        SKPaint eyePaint, SKPaint eyeHighlight, SKPaint mouthPaint, SKPaint blushPaint)
    {
        // Big sparkly eyes
        canvas.DrawCircle(center.X - eyeSpacing, eyeY, eyeRadius * 1.3f, eyePaint);
        canvas.DrawCircle(center.X + eyeSpacing, eyeY, eyeRadius * 1.3f, eyePaint);
        // Star-like highlights
        canvas.DrawCircle(center.X - eyeSpacing + eyeRadius * 0.3f, eyeY - eyeRadius * 0.4f,
            eyeRadius * 0.45f, eyeHighlight);
        canvas.DrawCircle(center.X + eyeSpacing + eyeRadius * 0.3f, eyeY - eyeRadius * 0.4f,
            eyeRadius * 0.45f, eyeHighlight);
        canvas.DrawCircle(center.X - eyeSpacing - eyeRadius * 0.3f, eyeY + eyeRadius * 0.2f,
            eyeRadius * 0.2f, eyeHighlight);
        canvas.DrawCircle(center.X + eyeSpacing - eyeRadius * 0.3f, eyeY + eyeRadius * 0.2f,
            eyeRadius * 0.2f, eyeHighlight);
        // Open mouth (O shape)
        canvas.DrawOval(center.X, center.Y + radius * 0.18f,
            radius * 0.1f, radius * 0.12f, mouthPaint);
        // Extra blush
        blushPaint.Color = new SKColor(255, 130, 130, 70);
        canvas.DrawCircle(center.X - radius * 0.35f, center.Y + radius * 0.1f,
            radius * 0.12f, blushPaint);
        canvas.DrawCircle(center.X + radius * 0.35f, center.Y + radius * 0.1f,
            radius * 0.12f, blushPaint);
    }

    private static void DrawContentFace(SKCanvas canvas, SKPoint center, float radius,
        float eyeY, float eyeSpacing, float eyeRadius,
        SKPaint eyePaint, SKPaint eyeHighlight, SKPaint mouthPaint)
    {
        // Relaxed eyes
        canvas.DrawCircle(center.X - eyeSpacing, eyeY, eyeRadius * 0.9f, eyePaint);
        canvas.DrawCircle(center.X + eyeSpacing, eyeY, eyeRadius * 0.9f, eyePaint);
        canvas.DrawCircle(center.X - eyeSpacing + eyeRadius * 0.3f, eyeY - eyeRadius * 0.3f,
            eyeRadius * 0.3f, eyeHighlight);
        canvas.DrawCircle(center.X + eyeSpacing + eyeRadius * 0.3f, eyeY - eyeRadius * 0.3f,
            eyeRadius * 0.3f, eyeHighlight);
        // Gentle smile
        using var gentleSmile = new SKPath();
        gentleSmile.MoveTo(center.X - radius * 0.15f, center.Y + radius * 0.12f);
        gentleSmile.QuadTo(center.X, center.Y + radius * 0.22f,
            center.X + radius * 0.15f, center.Y + radius * 0.12f);
        canvas.DrawPath(gentleSmile, mouthPaint);
    }
}

File: web/js/renderers/faceRenderer.js
import { PhaseConstants } from '../models/orbitalState.js';
import { drawLabel } from './sphereRenderer.js';

// Returns a face expression name based on moon phase angle.
export function getMoonFaceExpression(moonAngleDeg) {
  const angle = ((moonAngleDeg % 360) + 360) % 360;
  if (angle < PhaseConstants.NewMoonEnd) return 'sleepy';
  if (angle < PhaseConstants.WaxingCrescentEnd) return 'content';
  if (angle < PhaseConstants.FirstQuarterEnd) return 'happy';
  if (angle < PhaseConstants.WaxingGibbousEnd) return 'happy';
  if (angle < PhaseConstants.FullMoonEnd) return 'wow';
  if (angle < PhaseConstants.WaningGibbousEnd) return 'content';
  if (angle < PhaseConstants.LastQuarterEnd) return 'happy';
  if (angle < PhaseConstants.WaningCrescentEnd) return 'sleepy';
  return 'sleepy';
}

// Draws a cute cartoon face on a celestial body.
export function drawCuteFace(ctx, cx, cy, radius, expression) {
  const eyeY = cy - radius * 0.12;
  const eyeSpacing = radius * 0.25;
  const eyeRadius = radius * 0.07;

  const eyeColor = 'rgba(40,40,40,0.78)';
  const eyeHighlightColor = 'rgba(255,255,255,0.78)';
  const mouthWidth = radius * 0.04;
  const mouthColor = 'rgba(40,40,40,0.71)';
  const blushColor = 'rgba(255,130,130,0.2)';

  switch (expression) {
    case 'happy':
      drawHappyFace(ctx, cx, cy, radius, eyeY, eyeSpacing, eyeRadius,
        eyeColor, eyeHighlightColor, mouthWidth, mouthColor, blushColor);
      break;
    case 'sleepy':
      drawSleepyFace(ctx, cx, cy, radius, eyeY, eyeSpacing, eyeRadius,
        mouthWidth, mouthColor);
      break;
    case 'wow':
      drawWowFace(ctx, cx, cy, radius, eyeY, eyeSpacing, eyeRadius,
        eyeColor, eyeHighlightColor, mouthWidth, mouthColor, blushColor);
      break;
    case 'content':
      drawContentFace(ctx, cx, cy, radius, eyeY, eyeSpacing, eyeRadius,
        eyeColor, eyeHighlightColor, mouthWidth, mouthColor);
      break;
  }
}

function fillCircle(ctx, x, y, r, color) {
  ctx.fillStyle = color;
  ctx.beginPath();
  ctx.arc(x, y, r, 0, Math.PI * 2);
  ctx.fill();
}

function drawHappyFace(ctx, cx, cy, radius, eyeY, eyeSpacing, eyeRadius,
  eyeColor, eyeHighlightColor, mouthWidth, mouthColor, blushColor) {
  // Eyes
  fillCircle(ctx, cx - eyeSpacing, eyeY, eyeRadius, eyeColor);
  fillCircle(ctx, cx + eyeSpacing, eyeY, eyeRadius, eyeColor);
  // Eye highlights
  fillCircle(ctx, cx - eyeSpacing + eyeRadius * 0.3, eyeY - eyeRadius * 0.3,
    eyeRadius * 0.35, eyeHighlightColor);
  fillCircle(ctx, cx + eyeSpacing + eyeRadius * 0.3, eyeY - eyeRadius * 0.3,
    eyeRadius * 0.35, eyeHighlightColor);
  // Smile
  ctx.strokeStyle = mouthColor;
  ctx.lineWidth = mouthWidth;
  ctx.lineCap = 'round';
  ctx.beginPath();
  ctx.moveTo(cx - radius * 0.2, cy + radius * 0.1);
  ctx.quadraticCurveTo(cx, cy + radius * 0.3, cx + radius * 0.2, cy + radius * 0.1);
  ctx.stroke();
  // Blush cheeks
  fillCircle(ctx, cx - radius * 0.35, cy + radius * 0.08, radius * 0.1, blushColor);
  fillCircle(ctx, cx + radius * 0.35, cy + radius * 0.08, radius * 0.1, blushColor);
}

function drawSleepyFace(ctx, cx, cy, radius, eyeY, eyeSpacing, eyeRadius,
  mouthWidth, mouthColor) {
  ctx.strokeStyle = mouthColor;
  ctx.lineWidth = mouthWidth;
  ctx.lineCap = 'round';

  // Closed eyes (arcs)
  ctx.beginPath();
  ctx.moveTo(cx - eyeSpacing - eyeRadius, eyeY);
  ctx.quadraticCurveTo(cx - eyeSpacing, eyeY + eyeRadius * 0.8,
    cx - eyeSpacing + eyeRadius, eyeY);
  ctx.stroke();

  ctx.beginPath();
  ctx.moveTo(cx + eyeSpacing - eyeRadius, eyeY);
  ctx.quadraticCurveTo(cx + eyeSpacing, eyeY + eyeRadius * 0.8,
    cx + eyeSpacing + eyeRadius, eyeY);
  ctx.stroke();

  // Gentle smile
  ctx.beginPath();
  ctx.moveTo(cx - radius * 0.1, cy + radius * 0.15);
  ctx.quadraticCurveTo(cx, cy + radius * 0.22, cx + radius * 0.1, cy + radius * 0.15);
  ctx.stroke();

  // Zzz
  drawLabel(ctx, 'z z',
    cx + radius * 0.5, cy - radius * 0.3,
    radius * 0.2, [200, 200, 255, 150]);
}

function drawWowFace(ctx, cx, cy, radius, eyeY, eyeSpacing, eyeRadius,
  eyeColor, eyeHighlightColor, mouthWidth, mouthColor, blushColor) {
  // Big sparkly eyes
  fillCircle(ctx, cx - eyeSpacing, eyeY, eyeRadius * 1.3, eyeColor);
  fillCircle(ctx, cx + eyeSpacing, eyeY, eyeRadius * 1.3, eyeColor);
  // Star-like highlights
  fillCircle(ctx, cx - eyeSpacing + eyeRadius * 0.3, eyeY - eyeRadius * 0.4,
    eyeRadius * 0.45, eyeHighlightColor);
  fillCircle(ctx, cx + eyeSpacing + eyeRadius * 0.3, eyeY - eyeRadius * 0.4,
    eyeRadius * 0.45, eyeHighlightColor);
  fillCircle(ctx, cx - eyeSpacing - eyeRadius * 0.3, eyeY + eyeRadius * 0.2,
    eyeRadius * 0.2, eyeHighlightColor);
  fillCircle(ctx, cx + eyeSpacing - eyeRadius * 0.3, eyeY + eyeRadius * 0.2,
    eyeRadius * 0.2, eyeHighlightColor);
  // Open mouth (O shape)
  ctx.strokeStyle = mouthColor;
  ctx.lineWidth = mouthWidth;
  ctx.beginPath();
  ctx.ellipse(cx, cy + radius * 0.18, radius * 0.1, radius * 0.12, 0, 0, Math.PI * 2);
  ctx.stroke();
  // Extra blush
  const strongBlush = 'rgba(255,130,130,0.27)';
  fillCircle(ctx, cx - radius * 0.35, cy + radius * 0.1, radius * 0.12, strongBlush);
  fillCircle(ctx, cx + radius * 0.35, cy + radius * 0.1, radius * 0.12, strongBlush);
}

function drawContentFace(ctx, cx, cy, radius, eyeY, eyeSpacing, eyeRadius,
  eyeColor, eyeHighlightColor, mouthWidth, mouthColor) {
  // Relaxed eyes
  fillCircle(ctx, cx - eyeSpacing, eyeY, eyeRadius * 0.9, eyeColor);
  fillCircle(ctx, cx + eyeSpacing, eyeY, eyeRadius * 0.9, eyeColor);
  fillCircle(ctx, cx - eyeSpacing + eyeRadius * 0.3, eyeY - eyeRadius * 0.3,
    eyeRadius * 0.3, eyeHighlightColor);
  fillCircle(ctx, cx + eyeSpacing + eyeRadius * 0.3, eyeY - eyeRadius * 0.3,
    eyeRadius * 0.3, eyeHighlightColor);
  // Gentle smile
  ctx.strokeStyle = mouthColor;
  ctx.lineWidth = mouthWidth;
  ctx.lineCap = 'round';
  ctx.beginPath();
  ctx.moveTo(cx - radius * 0.15, cy + radius * 0.12);
  ctx.quadraticCurveTo(cx, cy + radius * 0.22, cx + radius * 0.15, cy + radius * 0.12);
  ctx.stroke();
}
